### PR TITLE
Add V3 keystone support, expose configuration options for domain and project

### PR DESCRIPTION
**Pull request by LarsFronius**
This was based on @imduffy15 s work and rebased against the 2.8 version of the plugin.

### CODE REVIEW

File: src/main/java/jenkins/plugins/openstack/compute/JCloudsCloud.java [hunk 1]
```diff
@@ -65,10 +65,14 @@
...
 
-    public final @Nonnull String endPointUrl;
     public final @Nonnull String identity;
     public final @Nonnull Secret credential;
+    public final @Nonnull String endPointUrl;
+    public final @Nonnull String project;
+    public final String domain;
+    public final String region;
     public final String zone;
+    public final int scriptTimeout;
```
Comment (olivergondza) [inline src/main/java/jenkins/plugins/openstack/compute/JCloudsCloud.java]: This seems reintroduced accidentally.

File: src/main/java/jenkins/plugins/openstack/compute/JCloudsCloud.java [hunk 2]
```diff
@@ -91,7 +95,7 @@ public ComputerLauncher createLauncher(@Nonnull JCloudsSlave slave) throws IOExc
                 SlaveOptions opts = slave.getSlaveOptions();
                 String credentialsId = opts.getCredentialsId();
                 if (credentialsId == null) {
-                    throw new ProvisioningFailedException("No ssh credentials selected");
+                    throw new ProvisioningFailedException("No ssh credentials selected", new Throwable());
```
Comment (olivergondza) [inline src/main/java/jenkins/plugins/openstack/compute/JCloudsCloud.java]: Why?

**Comment by olivergondza** [comment]
I will have to look at this in more detail. It is necessary to ensure this works with both versions of keystone and users will understand how to configure that correctly in both cases.

I do not like how the method signatures are bloating so I consider introducing a wrapper object for login details.

File: src/main/java/jenkins/plugins/openstack/compute/JCloudsCloud.java (hunk 1 above)
Comment (LarsFronius) [inline src/main/java/jenkins/plugins/openstack/compute/JCloudsCloud.java]: Removed it again 👍

**Comment by LarsFronius** [comment]
I agree 100% we should make sure this works on both versions. A wrapper object seems like a way to go. Is there a way for us to automatically test the wrapper?

**Comment by olivergondza** [comment]
I am running [ATH](https://github.com/jenkinsci/acceptance-test-harness/blob/master/src/test/java/plugins/OpenstackCloudPluginTest.java) that exercises Keystone v2. One way would be to ask someone like you, who have an access to keystone v3 instance, to run it against their instance and report problems.

An alternative would be to resurrect #85 and create integration tests that interact with openstack in docker container which would hopefully give as the flexibility to choose component versions.

**Comment by fraz3alpha** [comment]
I have access to an Openstack instance running v2 and v3 endpoints, and I've had a go at running the plugin with the V3 changes from this pull request merged in. 
The good news is that it will authenticate and then allows VM provisions to take place, but using the `client = OSFactory.builderV3()...` code doesn't allow for fallback to V2 authentication when a v2 API endpoint is used.
I don't know the best way to choose between the two builder methods - but falling back by catching the exception does at least functionally work as per the change in PR #83 (although I think maybe the exception message could be a little confusing if they both fail). Perhaps we could switch on the URL if we could guarantee that the strings v2 and v3 will occur in the API endpoint URL?

**Comment by LarsFronius** [comment]
Thanks for your input @fraz3alpha 
We already had a URL deployed as far as I know without "v3" in the URL.
I think we should go with a simple dropdown letting you choose whether it's a v3 or v2 the plugin should run against? This way its explicit and should only be very little prone to errors and implicit behaviour. I think I could manage plugging this in. The builder code from #83 for v2 works as is for you? I think I could manage to plug the dropdown and both builders based on it together if we can all agree this is the way to go.

**Comment by olivergondza** [comment]
> Is there a way for us to automatically test the wrapper?

I am working on a credentials wrapper and its AJAX passing to Descriptor method without so much magic. I will have something to build on hopefully soon.

**Comment by olivergondza** [comment]
What is the relation between auth field for v2 (endPointUrl, identity, credential, zone?) vs. v3(???)? Is v3 a superset of v2? IIUC project and domain are new, region is just renamed zone, right?

**Comment by fraz3alpha** [comment]
@LarsFronius : There was a couple of different variables used in your version vs. the other pull request, so they don't match up 1:1, but for reference I have been using this in my testing:

```
+        // Try version 2
+        try {
+            LOGGER.fine("Attempting to use V2 API endpoint");
+            return  OSFactory.builder().endpoint(endPointUrl)
+                    .credentials(identity, credential.getPlainText())
+                    .tenantName(project)
+                    .authenticate()
+                    .useRegion(region);
+        } catch (ClientResponseException e) {
+            // Failing that, try version 3
+            LOGGER.fine("Attempting to use V3 API endpoint");
+            return OSFactory.builderV3().endpoint(endPointUrl)
+                    .credentials(identity, credential.getPlainText(), Identifier.byName(domain))
+                    .scopeToProject(Identifier.byName(project), Identifier.byName(domain))
+                    .authenticate()
+                    .useRegion(region);
+        }
```
It also includes an extra domain parameter to credentials in the V3 version over what you had originally - I need to go back and check and see if that's actually required if you include the domain in the scopeToProject() call

Regarding how to differentiate between V2 and V3, I like your idea for a switchable set of parameters to make it explicit. Also, I see from v3.0.0 of openstack4j that they appear to have made some breaking changes that mean we may have to split out the V2 and V3 clients differently anyway if/when this plugin moves up to that version: http://www.openstack4j.com/learn/getting-started/#authenticate

**Comment by iamanikeev** [comment]
Hi! I've tested this on our v3 instance, works well. I had to add a region parameter to the plugin configuration, but this might be caused by the fact that I was using VERY old version previously, which didn't require region. Could you please let me know, what is required to merge this? Explicit distinction of v2 and v3 via drop-down option?

**Comment by olivergondza** [comment]
The problem is this breaks v2 support. Or at least was last time I checked.

**Comment by olivergondza** [comment]
Implemented in #133.

**Pull request closed without merging**